relay-server/src/actors/spooler/mod.rs

```diff
@@ -684,6 +697,55 @@ impl OnDisk {
             relay_statsd::metric!(gauge(RelayGauges::BufferEnvelopesDiskCount) = *count);
         }
     }
+
+    fn extract_key(&self, row: SqliteRow) -> Option<QueueKey> {
+        let own_key = row
+            .try_get("own_key")
+            .map_err(BufferError::FetchFailed)
+            .and_then(|key| ProjectKey::parse(key).map_err(BufferError::ParseProjectKeyFailed));
+        let sampling_key = row
+            .try_get("sampling_key")
+            .map_err(BufferError::FetchFailed)
+            .and_then(|key| ProjectKey::parse(key).map_err(BufferError::ParseProjectKeyFailed));
+
+        match (own_key, sampling_key) {
+            (Ok(own_key), Ok(sampling_key)) => Some(QueueKey {
+                own_key,
+                sampling_key,
+            }),
+            // Report the first found error.
+            (Err(err), _) | (_, Err(err)) => {
+                relay_log::error!("Failed to extract a queue key from the spool record: {err}");
+                None
+            }
+        }
+    }
+
+    /// Returns the index from the on-disk spool.
+    async fn get_spooled_index(
+        &self,
+    ) -> Result<BTreeMap<ProjectKey, BTreeSet<QueueKey>>, BufferError> {
+        let keys = sql::get_keys()
```

Comment:
Some time ago we discussed how querying the whole database during startup could be problematic for Relay. Is this no longer the case? Are there new considerations we've made?

Reply:
I've changed this a little bit. We do not block the start up , and execute the query in the background, meaning we allow ourself to run it a bit longer and process the result async.